crates/dyn-abi/src/eip712/parser.rs

```diff
@@ -139,13 +141,59 @@ impl<'a> EncodeType<'a> {
 
         Ok(Self { types })
     }
+
+    /// Computes the canonical string representation of the type.
+    ///
+    /// Orders the `ComponentTypes` based on the EIP-712 rules, removes unsupported whitespaces, and
+    /// validates them.
+    pub fn canonicalize(&self) -> Result<String, Error> {
+        // Ensure no unintended whitespaces
+        let mut resolver = Resolver::default();
+        for component_type in self.types.iter() {
+            resolver.ingest(component_type.to_owned());
+        }
+
+        // Resolve non-dependent types and validate them
+        let primary = resolver.non_dependent_types();
+        if primary.len() != 1 {
+            let msg = if primary.is_empty() {
+                "primary component".into()
+            } else {
+                format!(
+                    "primary component: {}",
+                    primary.iter().map(|t| t.type_name()).collect::<Vec<_>>().join(", ")
+                )
+            };
```

Comment:
you could use a mut string then push the type name if any, but imo this is also okay